store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java

```diff
@@ -71,10 +77,42 @@ public void delete() throws IOException {
 
     public RemoteFolder renameTo(String newName) throws IOException {
         info("renaming folder to '%s'", newName);
-        connection.rename(path, newName);
+
+        // ZBUG-2192
+        // execute only when last folder name's are different
+        // don't change any parent folder names
+        // using "path" as base name we are forming new folder name in such a way only last folder get's renamed at a time
+        
+        if (!isSameFolderName(newName)) {
+            String baseNewPath = path;
+            if (StringUtils.substringAfterLast(baseNewPath, PATH_SEPARATOR).isEmpty()
+                    && StringUtils.substringAfterLast(newName, PATH_SEPARATOR).isEmpty()) {
+                baseNewPath = newName;
+            } else {
+                baseNewPath = StringUtils.substringBeforeLast(baseNewPath, PATH_SEPARATOR) + PATH_SEPARATOR
+                        + StringUtils.substringAfterLast(newName, PATH_SEPARATOR);
+                if (StringUtils.endsWith(baseNewPath, PATH_SEPARATOR) || StringUtils.startsWith(baseNewPath, PATH_SEPARATOR)) {
+                    baseNewPath = StringUtils.removeEnd(baseNewPath, PATH_SEPARATOR);
+                    baseNewPath = StringUtils.removeStart(baseNewPath, PATH_SEPARATOR);
+                }
+            }
```

Comment:
can you let me know examples of path and newName which meet the condition at line 88, 91 and 94?

Reply:
Below are the scenarios for specific condition

- 88 ->When the path is “Test” and newName is “Test_1” (Basically single layer folder name which is modifiable)

- 91 -> When the path is “Test/a/pq” and newName is “Test/a_1/pq_1” (Multi-Layer Folder)

-  94 -> This is basically a check to prevent “/“ getting appended for baseNewPath before or after in any case, check for any unprecedented scenario

Reply:
Thank you for the scenarios. I understood.